#284 improve Emitter

Emily-Jiang opened the pull request:
Signed-off-by: Emily Jiang <emijiang6@googlemail.com>

Emily-Jiang commented:
@cescoffier can you take a look to see why the tests are failing? What is the completion criteria for the CompletionStage to complete?

cescoffier commented on smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java:
```diff
@@ -67,18 +73,18 @@ boolean isConnected() {
         if (msg instanceof Message) {
             //noinspection unchecked
-            emitter.onNext((Message) msg);
+            return CompletableFuture.runAsync(() -> emitter.onNext((Message) msg));
```
That's not correct.

You need something like (be aware this is pseudo-code):

```
CompletableFuture future = new CompletableFuture();
if (msg instanceof Message) {
    // here things need to be discussed
    // if the user passes a Message, he can attach the acknowledgment logic
   // however we cannot really do anything. We would need a way to customize the Supplier
} else {
   emitter.onNext(Message.of(msg, () -> future.complete(null));
}
```

Emily-Jiang replied:
updated

cescoffier commented on smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java:
```diff
@@ -268,13 +274,15 @@ public boolean isCaught() {
 
         public void run() {
-            emitter.send("a").send("b");
+            emitter.send("a");
```
The tests need to check that the returned CompletionStage is completed when the message is processed.

Emily-Jiang replied:
added more tests

cescoffier commented on smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/Emitter.java:
```diff
@@ -11,27 +13,28 @@
-     * @return the current emitter
-     * @throws IllegalStateException if the stream does not have any pending requests, or if the stream has been
-     *         cancelled or terminated.
+     * @return the {@code CompletionStage}
```
it needs to explain when the CS is completed.

Emily-Jiang replied:
done

cescoffier commented on smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/OnOverflow.java:
```diff
@@ -35,10 +35,17 @@
         BUFFER,
+        /**
+         * Buffers <strong>all</strong> values until the downstream consumes it.
+         * This creates an unbound buffer. If the buffer is full, an error is propagated.
```
_If the buffer is full_... in this case if the buffer is full, you are running out of memory :-)

Emily-Jiang replied:
done

cescoffier commented:
There is a design issue when the user passes a `Message`, as we cannot return a `CompletionStage` completed when the message is acknowledged.

One way would be to have 2 methods:

* `void send(Message<? extends T> message);` // can also return `this`
* `CompletionStage<Void> send(T payload);` 

We need to check how practical is it.

cescoffier commented:
For the record, the attached patch should fix the test failures and fix the generic signature.
[emitter.patch.txt](https://github.com/smallrye/smallrye-reactive-messaging/files/3868391/emitter.patch.txt)

cescoffier commented:
Remaining tasks:

- [ ] Test with a custom message implementation
- [ ] Test that the ack is correctly propagated to the returned CompletionStage.

Emily-Jiang commented:
@cescoffier please check whether you have more comments.

cescoffier commented:
The Kafka tests are failing because they use an emitter.
Apply the following diff:

```
         @Inject
         @Channel("temperature-values")
-        private Emitter<KafkaMessage<String, String>> emitter;
+        private Emitter<String> emitter;
```

Also, it would be great to rebase and cleanup the history a bit.

cescoffier commented:
We need more tests:
1. Test emitter with messages
2. Test that the acknowledgment is propagated correctly. This set of tests is very important because it's the core of this PR: the completion stage returned by the emitter but be completed when the message is acknowledged.

cescoffier commented:
Also, this PR is a breaking change on the API, so, after discussing with @kenfinnigan we would need to:

* keep the previous behavior for the Emitter class contained in the provider module (the old one, already deprecated)
* use the new API for the Emitter class from the `api` package.

Emily-Jiang commented:
I have added the tests to check the acknowledgment was propagated correctly though.

cescoffier commented on smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java:
```diff
@@ -137,21 +167,24 @@ public void testWithTwoEmittersConnectedToOneProcessor() {
     public static class MyBeanEmittingPayloadsUsingStream {
         @Inject
-        @Stream("foo")
+        @Channel("foo")
```
This test is about testing the legacy Stream annotation, do don't update to @Channel.

Emily-Jiang replied:
done

cescoffier commented on smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java:
```diff
@@ -220,16 +380,16 @@ public void run() {
     public static class BeanWithMissingStream {
         @Inject
-        @Stream("missing")
+        @Channel("missing")
```
This test is testing the behavior of the legacy @Stream annotation, so don't replace it.

Emily-Jiang replied:
done

Emily-Jiang commented:
@cescoffier I tried to add tests to test the deprecated Emitter. However, all plugging was removed by the package refractoring and the new ground needs to be laid out. I think if you want to rewire the old api, all of the scaffolding needs to be done again. This should be done under a different issue. By the way, I've addressed all of your comments except this one.

cescoffier commented:
Starting working on it to include the PR _soonish_.

cescoffier commented:
This PR is superseded by #339

Emily-Jiang commented:
Close this one because #339 superseded this.
